MapboxNavigationUI/RouteMapViewController.swift

```diff
@@ -242,41 +242,33 @@ extension RouteMapViewController: NavigationMapViewDelegate {
         
         let defaultReturn = CLLocation(coordinate: newCoordinate, altitude: location.altitude, horizontalAccuracy: location.horizontalAccuracy, verticalAccuracy: location.verticalAccuracy, course: location.course, speed: location.speed, timestamp: location.timestamp)
         
-        guard location.course == -1 else {
+        guard location.course != -1 else {
             return defaultReturn
         }
         
-        guard let coords = routeController.routeProgress.route.coordinates else {
-            return defaultReturn
-        }
+        let coords = routeController.routeProgress.currentLegProgress.priorCurrentUpcomingStepCoordinates
+        assert(!coords.isEmpty)
         
-        guard let closest = closestCoordinate(on: coords, to: location.coordinate) else {
-            return defaultReturn
-        }
+        let closest = closestCoordinate(on: coords, to: location.coordinate)!
         let slicedLine = polyline(along: coords, from: closest.coordinate, to: coords.last)
         
         // Get closest point infront of user
-        guard let infrontPointOne = coordinate(at: 2, fromStartOf: slicedLine) else {
-            return defaultReturn
-        }
-        guard let infrontCloest = closestCoordinate(on: coords, to: infrontPointOne) else {
-            return defaultReturn
-        }
+        let infrontPointOne = coordinate(at: 2, fromStartOf: slicedLine)!
```

Comment:
This code slices the line in order to get the coordinate two meters ahead of the snapped coordinate along the line. There is no difference between this code and `closest.coordinate.direction(to: coords[closest.coordinate.index + 1])`, except that the code here has to iterate over a potentially long step geometry.

Furthermore, you’d be hard-pressed to find a public road with two nodes spaced only two meters apart. A windy residential road may have two nodes spaced around ten meters apart, while a freeway would have two nodes spaced at least 20 meters apart. Anything closer than that is redundant and useless to us. (It’s called “overnoding” among OpenStreetMap mappers.)

What really matters is not an absolute distance but how far the user will travel in a fixed amount of time when going the speed limit. Roads with higher speed limits are built with gentler, broader curves. mapbox/MapboxDirections.swift#118 will give us something similar to speed limits. In the meantime, let’s decide on a time threshold like one second. Then divide the step’s distance by its expected travel time and multiply by the time threshold. Finally, pass that distance into `coordinate(at:fromStartOf:)`.

If that sounds familiar, it’s because we’ve done exactly this before, in #190.